### ISSUE TITLE
Install failed on "Generate Drupal project with composer package" with Ubuntu 16.04

### ISSUE THREAD

**Issue by danielrose28** [geerlingguy/drupal-vm#1460]
### Issue Type

 - Bug Report / Support Request

### Your Environment

```
Vagrant 1.9.5
VirtualBox 5.1.22r115126
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

### Your OS

 - **Linux** (Ubuntu 16.04)

### Full console output

![selection_025](https://user-images.githubusercontent.com/12372245/27743301-84d90180-5d89-11e7-8379-790d73ddb459.png)

Sorry, I couldn't scroll back to copy/paste everything, but I had this screen shot of the error.

### Summary

I literally just cloned the drupal-vm repo into my home directory. From within the drupal-vm folder I ran `vagrant up`. After a while it failed on "Generate Drupal project with composer package". I have the latest version of composer and it's up to date: `Composer version 1.4.2`. I believe everything else is up to date. I've seen some similar issues that were already closed with the fixes now in the master branch, but I'm still getting this error. 

Thanks for any help.

**Comment by geerlingguy** [comment]
It looks like the problem is that the directory /var/www/drupalvm/drupal/ is not empty—if you had tried building it before, or copied it from a place where drupal had already been installed before, Composer will refuse to create a new `drupal` directory and project so as to not wipe out the existing one.

If this is a new build, just delete the `drupal` directory inside the Drupal VM directory, and then run `vagrant provision` again (or `vagrant up` to kick off a new build if you deleted the existing build).

**Comment by danielrose28** [comment]
Ah, that makes sense. I'll give it a shot and see how it goes. Thanks Jeff.

**Comment by danielrose28** [comment]
I'm still actually getting this error. When I try to remove `drupal` it just says no such file or directory.

**Comment by danielrose28** [comment]
I spoke too soon. I totally did the wrong thing. I just deleted the drupal folder inside the dvm folder now so we'll see how it goes.

**Comment by danielrose28** [comment]
I'm actually finding that every other `vagrant up` and/or provision, it's crashing my computer. First couple of times I thought it was something else that was crashing my computer, but turns out, it is drupalvm. I get this error as I share above, every other install. While the other times it just crashes my computer. I've even added a local.config.yml and adjusted the memory, but it still crashes. When it does crash I am only able to catch a glimpse of the error, which is extremely long and then the exit error is always something about vboxmanage not found in path. However, when I run `vboxmanage -v` from my path I get `5.1.22r115126` as the current version. 

So to summarize.. this error that I've reported is fixed by deleting the `drupal` folder from the project. But then when I run `vagrant up --provision` it runs for a long time and then fails with the vboxmanage error, while also crashing my computer every time. I can try a couple of other ways, like after I delete the drupal folder, maybe run `vagrant halt && vagrant destroy` and then `vagrant up` again or `vagrant up --provision`, if it matters.

**Comment by oxyc** [comment]
Did you happen to notie on what task, or close to what task it crashes?

**Comment by oxyc** [comment]
You can redirect the output to a file so you can inspect it after the crash btw.

```
vagrant provision 2>&1 | tee vagrant.log
```

**Comment by danielrose28** [comment]
I apologize. I haven't been able to determine what job makes it crash yet. I will definitely try your recommendations and try to get the output so I can share here in a gist. Thanks for the tips.

**Comment by danielrose28** [comment]
So I ran `vagrant up 2>&1 | tee vagrant.log` and output this to a file: https://gist.github.com/danielrose28/d5227416bd42da936b7448317df5f00f. You'll see that it does hang on "Generate Drupal project with composer package." This happens even when I delete the drupal directory from the drupalvm project folder and start over. There is some more exit output which I cannot grab, nor does it get saved to the file due to crashing. I believe the exit output refers to vboxmanage, but again, I have the latest versions of these dependencies, etc. I'm honestly not sure what to try next.

**Comment by oxyc** [comment]
Not sure what this could be but I'll list a few things worth trying

Worth checking `/var/log/syslog`, `/var/log/kern.log` for what the time when your host system crashed. Maybe worth checking the guest machine too if there are any clues.

---

> Jul 05 17:05:31 gazelle systemd[1]: Starting NFS server and services...
Jul 05 17:05:31 gazelle exportfs[1514]: exportfs: duplicated export entries:
Jul 05 17:05:31 gazelle exportfs[1514]: exportfs:         192.168.88.88:/home/daniel/drupal-vm
Jul 05 17:05:31 gazelle exportfs[1514]: exportfs:         192.168.88.88:/home/daniel/drupal-vm
Jul 05 17:05:31 gazelle systemd[1]: Started NFS server and services.
exportfs: duplicated export entries:
exportfs: 	192.168.88.88:/home/daniel/drupal-vm
exportfs: 	192.168.88.88:/home/daniel/drupal-vm

There's a duplicate entry in `/etc/exports`, if you edit the file you can remove both entries. `vagrant up` should add the correct one.

---

Try using `rsync` rather than NFS just to see if it's NFS related. I believe this is the first task that writes to a NFS shared location.

```yaml
vagrant_synced_folders:
  - local_path: .
    destination: /var/www/drupalvm
    type: rsync
    create: true
```

If this works, try switching back to NFS but with Drush make rather than the default `composer create-project`. http://docs.drupalvm.com/en/latest/deployment/drush-make/

Possibly related:
- https://www.virtualbox.org/ticket/14089

**Comment by danielrose28** [comment]
Forgot to mention that I also got an output into a file while running the `--debug` flag as well. Would that help? In the meantime, I'll try some more recommendations from above. Thanks again for your help with this.

**Comment by danielrose28** [comment]
So running with `vagrant_synced_folders` type set to `rsync` finished successfully. However, when I then switched back to NFS and ran with drush make, I got an [error](https://gist.github.com/danielrose28/c17e94f14994cdcf1bb18486acf6f8bd#file-failed-install-with-drush-make).

**Comment by oxyc** [comment]
As you're on linux, it's worth checking if it works with [NFSv4](http://docs.drupalvm.com/en/latest/getting-started/installation-linux/#nfs-shares-are-slow-or-hang-due-to-udp-protocol). Switch back to the default `composer create-project` installation and add:

```yaml
vagrant_synced_folders:
  # The default share.
  - local_path: .
    destination: /var/www/drupalvm
    type: nfs
    create: true
    options_override:
      nfs_version: 4
      nfs_udp: false

  # Manually specify the /vagrant share to override the defaults
  - local_path: .
    destination: /vagrant
    type: nfs
    options_override:
      nfs_version: 4
      nfs_udp: false
```

**Comment by danielrose28** [comment]
@oxyc do I just copy and paste the above example code exactly as is?

**Comment by oxyc** [comment]
Yes, replacing whatever you have as `vagrant_synced_folders` from before.

**Comment by danielrose28** [comment]
Well, I tried these and still got the same error. Crashed at same spot with same error.

**Comment by oxyc** [comment]
Just to be clear, the error is your system crashing, right? If it's the _directory /var/www/drupalvm/drupal/ is not empty_ you just need to delete the `drupal/` directory before provisioning.

If it's your system crashing, I'm running out of ideas. Earlier you confirmed:
- It works with NFS and drush make (the task that failed was something unrelated to this)
- It works with rsync

@geerlingguy any ideas?

To get around the problem, I would create the Drupal project on my host machine and let NFS sync it to the VM, rather than vice versa. I guess the system crash might still happen again on some other task though...

1. In the root of your project, next to the Vagrantfile create a drupal project, the same command as when the playbook errors out

        composer create-project drupal-composer/drupal-project:8.x-dev drupal --stability dev --no-interaction

2. Build the VM and the task should be skipped as the files are already in place.

        vagrant up

**Comment by danielrose28** [comment]
You are correct. Each time I've also deleted the drupal folder, restarted NFS server, ran `service ufw stop` and still get the system crash with the same composer related error. I appreciate all of your help. Sorry I'm not able to get this up and running. I'm happy to just give up. :(

**Comment by oxyc** [comment]
You could still build the Drupal core files on your host first. I do this just because it's way faster when provisoning.

Also, did you check if there were any clues in /var/log/syslog, /var/log/kern.log?

**Comment by danielrose28** [comment]
I took a look, but nothing stood out at me. I'll put them in a gist and share here. In the meantime, here is a [vagrant log with debug flag enabled](https://gist.github.com/danielrose28/28b4d8b43aae28caa4207e7bfdfbe916#file-vagrant-log) that I ran the other day as well.

**Comment by oxyc** [comment]
Apparently there are also virtualbox logs in `$HOME/.VirtualBox/Machines/<vm name>/Logs`, worth checking.

**Comment by skorasaurus** [comment]
I'm also experiencing this same error that @danielrose28. 

Ubuntu 16.04; Virtualbox 5.1.18-114002~Ubuntu~xenial; 

@geerlingguy - Regarding the hanging that you mention; 

the vagrant up process would indefinitely hang at: TASK [geerlingguy.drupal : Generate Drupal project with composer package.] *** however this would not completely crash my computer.

**Comment by oxyc** [comment]
Hmm, @skorasaurus you've been using Drupal VM for a while right? Were you using some other build preference before or did it used to work with `drupal_build_composer_project: true`? Wondering if it's some issue with a recent version of Vagrant or VirtualBox.

**Comment by skorasaurus** [comment]
@oxyc - Yes but I had last launched my VM about 3-4 weeks ago, initially received a different error, and just decided to start from scratch again (rm'ed my drupal-vm directory, recloned it, created a new VM etc). Same OS and hardware.

**Comment by oxyc** [comment]
At least [here](https://github.com/geerlingguy/drupal-vm/issues/1163) you stated that you used `build_composer_project` and the VM did build. I'm fairly sure this isn't something Drupal VM could have introduced since then so sounds like it could be a bug in Vagrant or VirtualBox since you filed that issue.

**Comment by skorasaurus** [comment]
yep, I may need to go back and check my logs in and /var/log/apt/history.log and revert some versions.

**Comment by oxyc** [comment]
Let us know if you find anything. Also, did you check `/.VirtualBox/Machines/<vm name>/Logs` and `/var/log/syslog` if there was anythign information about why it crashed?

**Comment by skorasaurus** [comment]
I first upgraded my extensions pack to 5.1.22 and first; rebooted; just in case to see if the error has been fixed since then; alas, it's not; still hung at 'Generate Drupal project with Composer package' but my computer is still responsive... 

Here's my Virtualbox log: https://gist.github.com/skorasaurus/e3117c413cb1c4d48520dedb6c64be86

and my syslog https://gist.github.com/skorasaurus/1f9cea059d6e20dac3c2249d6945b09a

**Comment by oxyc** [comment]
Thanks for the logs! Unfortunately I don't see what would be causing the crash :/.

The task, _Generate Drupal project with composer package_, where the system crashes is very intense IO-wise though, so probably has something to do with that. I'd recommend avoiding the task and generating the project manually on your host before provisioning. I do this on all my projects because it's way faster as well.

The Drupal VM installation steps would be:
1. Download and extract Drupal VM
2. Enter the directory
3. Generate the codebase in the expected `drupal/` directory:

       composer create-project drupal-composer/drupal-project:8.x-dev drupal --stability dev --no-interaction

4. Build the VM

       vagrant up

Alternatively any of the other build methods should work

**Comment by oxyc** [comment]
@skorasaurus just saw you said it _hung_ but didn't crash. How long did you leave it running?

**Comment by skorasaurus** [comment]
@oxyc ; I've let it run for as long as about 30-45 minutes.

**Comment by oxyc** [comment]
Seems too long for sure (also https://github.com/geerlingguy/drupal-vm/issues/1309), but I think that's a different issue.

Could you open up a new issue and mention what type of hardware you have as well. Do you have a hard drive or a SSD? Also post your `config.yml`. Did you try using [TCP rather than UDP](http://docs.drupalvm.com/en/latest/getting-started/installation-linux/#nfs-shares-are-slow-or-hang-due-to-udp-protocol) for NFSv4?

**Comment by skorasaurus** [comment]
@oxyc - using TCP instead of UDP in my config.yml worked for me; made a new issue on it (#1476)

**Comment by geerlingguy** [comment]
K, let's close this issue and focus on closing out #1476 — I'm going to be merging in @oxyc's upstream drupal role fix to install into tmp then copy into the final drupal installation directory.
